File: judo.java
import java.io.*;
import java.util.*;

public class judo{

    public static void main(String[] args){
        try{
            String fileName=args[0];
            String line=null;           //read the data
            FileReader fileReader=new FileReader(fileName);
            BufferedReader BufferedReader=new BufferedReader(fileReader);
            ArrayList<String> words=new ArrayList<String>();
            while((line=BufferedReader.readLine()) != null){
                if(line.trim().length()>0){  //ignore the empty lines and all puctuation etc, make everything lower case
                    for(String word : line.split("\\s+")){
                        words.add(word.replaceAll("[^a-zA-Z ]", "").toLowerCase());
                    }
                }
            }
            
            //build a unique word list
            Set<String> uniquewords = new LinkedHashSet<>(words);
            String[] uwords = uniquewords.toArray(new String[uniquewords.size()]);
 
            BufferedReader.close();
            
            Arrays.sort(uwords);  //sort and print to screen the alphabetised list
            for(int i=0;i<uwords.length;i++){
                System.out.println(uwords[i]);
            }
            
            ArrayList<String> all = new ArrayList<String>();
            all=make6(uwords);  //all the 6 letter words
            
            String[] perms=all.toArray(new String[all.size()]);
            
            ArrayList<String> asix=new ArrayList<String>();
            asix=filt(uwords,6);
            String[] six = asix.toArray(new String[asix.size()]);
            
            List x=Arrays.asList(perms);
            List y=Arrays.asList(six);

            System.out.println("Part 2:");

            for(int i=0;i<six.length;i++){  //which 6 letter word is a combination?
                if(x.contains(y.get(i))){
                    System.out.println(y.get(i));
                }
            }

            
        }
        
        catch(FileNotFoundException ex){
            System.out.println("File not found");
        }
        
        catch(IOException ex){
            System.out.println("Error reading file");
        }
        
        
    }
    
    public static ArrayList<String> filt(String[] words, int n){
        //function that takes an array of words and returns the subset of length n
        ArrayList<String> list=new ArrayList<String>();
        for(int i=0;i<words.length;i++){
            if(words[i].length()==n){
                list.add(words[i]);
            }
        }
        
        return list;
    }
    
    public static ArrayList<String> make6(String[] words){
        //function that takes an array of words and makes all possible combinations of length 6 - really ugly!
        ArrayList<String> aone=new ArrayList<String>();
        ArrayList<String> atwo=new ArrayList<String>();
        ArrayList<String> athree=new ArrayList<String>();
        ArrayList<String> afour=new ArrayList<String>();
        ArrayList<String> afive=new ArrayList<String>();
        
        aone=filt(words,1);
        atwo=filt(words,2);
        athree=filt(words,3);
        afour=filt(words,4);
        afive=filt(words,5);
        
        String[] one = aone.toArray(new String[aone.size()]);
        String[] two = atwo.toArray(new String[atwo.size()]);
        String[] three = athree.toArray(new String[athree.size()]);
        String[] four = afour.toArray(new String[afour.size()]);
        String[] five = afive.toArray(new String[afive.size()]);
        
        ArrayList<String> new_words=new ArrayList<String>();
        
        for(int i=0;i<one.length;i++){
            for(int j=0;j<five.length;j++){
                new_words.add(one[i]+five[j]);
                new_words.add(five[j]+one[i]);
            }
        }
        
        for(int i=0;i<two.length;i++){
            for(int j=0;j<four.length;j++){
                new_words.add(two[i]+four[j]);
                new_words.add(four[j]+two[i]);
            }
        }
        
        for(int i=0;i<three.length;i++){
            for(int j=0;j<three.length;j++){
                new_words.add(three[i]+three[j]);
            }
        }
        return new_words;
    }
}

File: judo.py
import numpy as np
import sys

def read_data(filename):  #read the data
    text=[]
    words=[]
    with open(filename,'r') as f:
        for line in f:
            l=line.replace('\xef\xbb\xbf','').rstrip('\n\r')   #strip all the formatting stuff
            if l != '':   #ignore the empty lines
                text.append(l)
                words.append(l.split(' '))   #split each sentence into words
    words=[i for j in words for i in j]  #flatten
    for i in range(len(words)):words[i]=words[i].strip('`"()!').strip("'").rstrip(',.!:;?)').lower() #remove all the punctuation and make everything lower case
    return text,words

def filt(words,n):
    #filter(lambda x:len(x)==n,words)
    return [x for x in words if len(x)==n]   #list of the words of length n

def make6(words):   #make all the 6 letter words
    one=filt(words,1)
    two=filt(words,2)
    three=filt(words,3) #ugly way to do this
    four=filt(words,4)
    five=filt(words,5)

    new_words=[]
    
    #six letter words are either 1+5,2+4,3+3 both way round
    #make a list of all possible 6 letter words

    for i in range(len(one)):
        new_words.append([s+one[i] for s in five])   #lists of all permuations
        new_words.append([one[i]+s for s in five])

    for i in range(len(two)):
        new_words.append([s+two[i] for s in four])
        new_words.append([two[i]+s for s in four])

    for i in range(len(three)):
        new_words.append([s+three[i] for s in three])

    new_words=[i for j in new_words for i in j]  #flatten

    return new_words

if __name__=="__main__":

    filename=sys.argv[1]
    print 'Reading text from file: ', filename

    try:
        data,words=read_data(filename)  #read the data
    except IOError as e:
        print "I/O error({0}): {1}".format(e.errno, e.strerror)

    unique_words=set(words)   #build a list of unique words
    au_words=sorted(unique_words)  #sort alphabetically
    nocc=[None]*len(au_words)
    for i in range(len(au_words)):nocc[i]=words.count(au_words[i])  #how many times does each word appear
    print 'There are %i unique words in the file\n'% (len(au_words))
    print 'Here is a list of all the unique words in the file alphabetically and the number of times they appear:\n'
    for i in range(len(au_words)): print au_words[i],' occurred ',nocc[i],' times'

    #all the 6 letter words are
    six=filt(au_words,6)

    #six letter words are either 1+5,2+4,3+3 both way round
    #make a list of all possible 6 letter words

    all=make6(au_words)

    print 'Part 2:'

    for i in range(len(six)):
        if all.count(six[i])==1:print six[i]  #does this 6 letter word appear in the list of all possible 6 letter words?




